### PR TITLE
Improvements/frontend responsibilities

### PR DESCRIPTION
**Pull request by einari**
The purpose of this pull request is to look at assigning responsibilities in a clearer way for the frontend.

Primary design principles:

- ViewModels deal with fetching models and communication both internally and externally
- Views observe ViewModels and never tell external parties to change, not even indirectly

## ViewModel lifecycle

The ViewModels have a different lifecycle than views. In the React model, any mutation in state will cause the View to re-render. The view must therefor take a passive role with regards to both the population of data coming into it and what it needs, but also what is communicated to other parties.

An example of a problem we had was that the Views where telling what to populate the portal structure directly. When we pop up a dialog - which is then being shown as a consequence from a state change (show dialog booleans), it communicated the change for every render.

The nature of the ViewModel is to expose behaviors and state - the state is wrapped in a way that automatically mutates the views when the state changes. So indirectly the ViewModel affects the lifecycle of the View. The direction of flow is correct. With the way we had it before, we told the state to change as a consequence of the view re-rendering. Which is a wrong direction of flow.

## Formalizations

This pull request brings with it a couple of declarative formalizations that can now be used in other microservices:

- Toolbar
- ActionBar
- Bootstraper

Both are formalized as components, but in reality they wrap the Toolbar and ActionBar APIs - but called from the ViewModels of these components.

## Whats missing?

There is one thing that is missing - bindings between components done at a composition level. 
As of now, the `<AssignMicroservice/>` component does not get the `selectedApplication` set - since this is now effectively disabled. We will have to return to this. A specific task is created for figuring out how to do this on a compositional level in a good way.

### CODE REVIEW

**Review by pavsaund — CHANGES_REQUESTED** [review]
I like how this is evolving, especially the declarative portal components. Really interesting to see how the Toolbar Items was solved as well.

Most comments are suggestions, except these which should be fixed: 
- Populate navigation items after query is complete.
- Rename ApplicationViewModel.tsx => *.ts
- Rename Bootstraper => Bootstrapper

Rest are suggestions.

File: Source/Shared/Components/toolbar/ToolbarProps.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+export interface ToolbarProps {
+    children?: JSX.Element[] | JSX.Element;
```
[the remark this answers is not in the document]
Comment (einari) [reply Source/Shared/Components/toolbar/ToolbarProps.ts:5]: Good point. Maybe we can even use the actual ToolbarItem React element we now have as well. Then it should limit to the actual type.

File: Source/Shared/Components/toolbar/ToolbarViewModel.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import { injectable } from 'tsyringe';
+
+@injectable()
+export class ToolbarViewModel {
+
+    constructor(private readonly _toolbarItems: ToolbarItems) {
+
+    }
+
+    activate(viewContext: IViewContext<ToolbarViewModel, ToolbarProps>) {
```
[the remark this answers is not in the document]
Comment (einari) [reply Source/Shared/Components/toolbar/ToolbarViewModel.ts:17]: Absolutely - a little bit out of context of this PR - I'll register a task for it.

File: Source/Shared/Components/toolbar/ToolbarProps.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavsaund) [reply Source/Shared/Components/toolbar/ToolbarProps.ts:5]: I tried that first without much luck. It complained in the view that the child needed to be of type `Element` and have 26 additional props... Hope I missed something obvious. Perhaps the component or its props needs to be of a certain generic base type to support the generic Element props?

**Comment by einari** [comment]
And good suggestions they were :)

File: Source/Shared/Components/toolbar/ToolbarProps.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (einari) [reply Source/Shared/Components/toolbar/ToolbarProps.ts]: needs some investigation - went with what you suggested.

**Review by pavsaund — APPROVED** [review]

**Comment by pavsaund** [comment]
How do we bump the version of this PR and deploy it as part of accepting it?

**Comment by pavsaund** [comment]
OK, so I think the version comes in to play when merging code into main. So I'll accept this now and deal with the consequences if that's the wrong way to go about things :D

**Pull request merged**